waspc/src/Wasp/AppSpec/Valid.hs

```diff
@@ -0,0 +1,91 @@
+{-# LANGUAGE TypeApplications #-}
+
+module Wasp.AppSpec.Valid
+  ( validateAppSpec,
+    ValidationError (..),
+    getApp,
+    isAuthEnabled,
+  )
+where
+
+import Data.List (find)
+import Data.Maybe (isJust)
+import Wasp.AppSpec (AppSpec)
+import qualified Wasp.AppSpec as AS
+import Wasp.AppSpec.App (App)
+import qualified Wasp.AppSpec.App as App
+import qualified Wasp.AppSpec.App.Auth as Auth
+import Wasp.AppSpec.Core.Decl (takeDecls)
+import qualified Wasp.AppSpec.Entity as Entity
+import qualified Wasp.AppSpec.Entity.Field as Entity.Field
+import qualified Wasp.AppSpec.Page as Page
+
+data ValidationError = GenericValidationError String
+  deriving (Show, Eq)
+
+validateAppSpec :: AppSpec -> [ValidationError]
+validateAppSpec spec =
+  case validateExactlyOneAppExists spec of
+    Just err -> [err]
+    Nothing ->
+      -- NOTE: We check these only if App exists because they all rely on it existing.
+      concat
+        [ validateAppAuthIsSetIfAnyPageRequiresAuth spec,
+          validateAuthUserEntityHasCorrectFields spec
+        ]
+
+validateExactlyOneAppExists :: AppSpec -> Maybe ValidationError
+validateExactlyOneAppExists spec =
+  case AS.takeDecls @App (AS.decls spec) of
+    [_] -> Nothing
+    apps ->
+      Just $
+        GenericValidationError $
+          "Expected exactly 1 'app' declaration, but found " ++ show (length apps) ++ "."
+
+validateAppAuthIsSetIfAnyPageRequiresAuth :: AppSpec -> [ValidationError]
+validateAppAuthIsSetIfAnyPageRequiresAuth spec =
+  if anyPageRequiresAuth && not (isAuthEnabled spec)
+    then
+      [ GenericValidationError
+          "Expected app.auth to be defined since there are Pages with authRequired set to true."
+      ]
+    else []
+  where
+    anyPageRequiresAuth = any (isJust . Page.authRequired) (snd <$> AS.getPages spec)
+
+validateAuthUserEntityHasCorrectFields :: AppSpec -> [ValidationError]
+validateAuthUserEntityHasCorrectFields spec = case App.auth (snd $ getApp spec) of
+  Nothing -> []
+  Just auth ->
+    let userEntity = snd $ AS.resolveRef spec (Auth.userEntity auth)
+        fields = Entity.getFields userEntity
+        maybeEmailField = find ((== "email") . Entity.Field.fieldName) fields
```

Comment:
does this have to be hardcoded?

Reply:
Ok so this is tricky!

So this is really a source of the truth -> in AppSpec, we make this requirement that userEntity must have email and password fields, if it is used in auth as userEntity and EmailAndPassword method is used.

So this is the only place where this is enforced, check or used so far.

Later, down the stream, Generator is the one that actually uses this info. Right now this is encoded directly in one of the Generator templates, the fact that Generator expects email and password field to exist, and that they are named `email` and `password`.

Now, I could extract `email` and `password` strings here in AppSpec into a variable/binding and then use that from Generator. That would be the way to remove the "hardcoding". However, there would still be a lot of assumptions there -> Generator assuming these are of correct type in the Entity, Generator assuming they exist, Generator merely knowing about them. I wonder if it will bring much value to extract them, it feels like we would need to do much more to establish this connection beyond assumptions.

I feel this is relatively complex for now, so I left a comment for now in AuthG.hs explaining that this is hardcoded and that we might want in the future to figure out a way to make this better encoded in our code.